letsencrypt/models/letsencrypt.py

```diff
@@ -2,167 +2,223 @@
+                    self._respond_challenge_dns(challenge, domain)
+                    client.answer_challenge(
+                        challenge, challenges.DNSResponse())
+                    auth_responded = True
+                    break
+        if not auth_responded:
+            raise exceptions.ValidationError(
+                _('Could not respond to letsencrypt challenges.'))
+        # let them know we are done and they should check
+        order_resource = client.poll_and_finalize(authzr)
+        with open(join(_get_data_dir(), '%s.crt' % domain), 'w') as crt:
+            crt.write(order_resource.fullchain_pem)
+            _logger.info('SUCCESS: Certificate saved :%s', crt.name)
+            ir_config_parameter = self.env['ir.config_parameter']
+            reload_cmd = ir_config_parameter.get_param(
+                'letsencrypt_reload_command', False)
+            if reload_cmd:
+                self._call_cmdline(['sh', '-c', reload_cmd])
+            else:
+                _logger.warning("No reload command defined.")
+
+    def _create_client(self, account_key):
+        net = client.ClientNetwork(account_key)
+        if self.env.context.get('dry_run'):
+            directory_url = V2_STAGING_DIRECTORY_URL
+        else:
+            directory_url = V2_STAGING_DIRECTORY_URL  # TODO
+        directory_json = requests.get(directory_url).json()
+        return client.ClientV2(directory_json, net)
+
+    def _make_csr(self, account_key, domain_key_file, domain):
+        parameter_model = self.env['ir.config_parameter']
+        altnames = parameter_model.search([
+            ('key', '=', 'letsencrypt_altnames')],
+        ).value
+        if altnames:
+            altnames = altnames.split(',')
+            valid_domains = altnames + [domain]
         else:
-            _logger.info('no command defined for reloading webserver, please '
-                         'do it manually in order to apply new certificate')
+            valid_domains = [domain]
+        _logger.info(
+            'Making CSR for the following domains: ' + str(valid_domains))
+        return crypto_util.make_csr(
+            open(domain_key_file).read(), valid_domains)
+
+    def _respond_challenge_http(self, challenge, account_key):
+        """
+        Respond to the HTTP challenge by writing the file to serve.
+        """
+        token = base64.urlsafe_b64encode(challenge.token)
+        challenge_file = join(_get_challenge_dir(), '%s' % token.rstrip('='))
+        with open(challenge_file, 'wb') as challenge_file:
+            challenge_file.write(token.rstrip('=') + '.' + jose.b64encode(
+                account_key.thumbprint(hash_function=hashes.SHA256)).decode())
+
+    def _respond_challenge_dns(self, challenge, domain):
+        """
+        Respond to the DNS challenge by creating the DNS record
+        on the provider.
+        """
+        ir_config_parameter = self.env['ir.config_parameter']
```

Comment:
I'd expect this function to look at the dns provider selected, and call `self._respond_challenge_dns_$provider` - this way, you can move the shell code to `_respond_challenge_dns_shell` and have nothing else to do.

Another way to do this would be that all other modules override this function, and only do something when their dns provider is selected. I personally prefer the first method.